### ISSUE TITLE
About results

### ISSUE THREAD

**Issue by ceasorcai** [huiqu18/FullNet-varCE#6]
Thanks for sharing the codes. I trained the code for the segmentation of The MICCAI 2015 gland segmentation challenge dataset (GlaS). The learning rate 0.0005, epochs 1000. The performance is worse than your paper. I do not know why, can you help me?

                                                     testA                 |              testB
                                        F1  | Dice_obj | Hausdoff | F1 | Dice_obj | Hausdoff
Paper( FullNet)             | 0.914  | 0.909    | 40.38     | 0.855  | 0.853    | 99.46
trained FullNet             | 0.8925 | 0.9033 | 41.9818 | 0.8148 | 0.8537 | 95.0118
Paper(FullNet&varCE)  | 0.924   | 0.914   | 37.28     | 0.853   | 0.856   | 88.75
trained FullNet&varCE | 0.9032 | 0.9064 | 43.9789 | 0.8450 | 0.8638 | 81.6378

**Comment by ceasorcai** [comment]
@huiqu18

**Comment by huiqu18** [comment]
Did you use ternary labels for training, i.e., gland inside area, gland contour, and background?

**Comment by MegumiOuO** [comment]
@huiqu18  i have tried to use eroded single channel label to replace ternary labels, however it didin‘t perform very well.  So, can you provide the code for generating ternary labels for training?

**Comment by thinkInJava33** [comment]
@huiqu18 can you share the pretrained model of your best result?

**Comment by Xu-bowen** [comment]
why i find some picture in GlaS testB have wrong label??testB3,11,18,19only have 1 unique label??

**Comment by huiqu18** [comment]
> @huiqu18 i have tried to use eroded single channel label to replace ternary labels, however it didin‘t perform very well. So, can you provide the code for generating ternary labels for training?

The code for generating ternary labels can be found in the newly added `prepare_data.py` file.

**Comment by huiqu18** [comment]
> @huiqu18 can you share the pretrained model of your best result?

Pretrained models of FullNet-varCE for both datasets can be downloaded from the links provided in updated README.

**Comment by huiqu18** [comment]
> why i find some picture in GlaS testB have wrong label??testB3,11,18,19only have 1 unique label??

I guess that the instances in each of those images are from one large malignant gland in the original pathology image. The authors cut the patch from that large gland, resulting in one unique label for those instances. We treated them as different glands because it is impossible to know this information for the model given the images in the dataset.

**Comment by MegumiOuO** [comment]
@huiqu18  i found the results of different experiments that using same setting will be different by 2~3%  even more on Hausdoff. Do you have this phenomenon？

**Comment by huiqu18** [comment]
> @huiqu18 i found the results of different experiments that using same setting will be different by 2~3% even more on Hausdoff. Do you have this phenomenon？

It is normal and caused by the random operations in weight initialization, data augmentation, batch sampling, cuda functions, etc. The variance could be reduced but will still exist given enough training data. The Hausdorff distance is sensitive to small noisy areas, thus its variance is relatively larger.

If you want to get the same results under the same setting, you can try to fix the random seed as follows:
```
seed = 1  # whatever you want
torch.manual_seed(seed)
torch.cuda.manual_seed(seed)
torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False
np.random.seed(seed)
random.seed(seed)
```
However, it is not recommended because it will reduce performance. And for the FullNet, fixing random seed will also cause much larger memory consumption, without using the optimizations in cudnn benchmark.

Here is a link about the reproducibility problem in pytorch: https://pytorch.org/docs/stable/notes/randomness.html.

**Comment by ceasorcai** [comment]
> Did you use ternary labels for training, i.e., gland inside area, gland contour, and background?

Thanks, I tried,but my performance still not as good as yours. I have another question, Can you tell me how to set the val dataset for gland segmentation challenge dataset (GlaS)? Thanks

**Comment by ceasorcai** [comment]
@huiqu18

**Comment by huiqu18** [comment]
> > Did you use ternary labels for training, i.e., gland inside area, gland contour, and background?
> 
> Thanks, I tried,but my performance still not as good as yours. I have another question, Can you tell me how to set the val dataset for gland segmentation challenge dataset (GlaS)? Thanks

In gland segmentation, we use the whole training set as train set and the testA set as validation set.

**Comment by ceasorcai** [comment]
Thanks
